File: scripts/generate-quiz-json.js
#!/usr/bin/env node
/**
 * Quiz JSON generator for Prisma seeding
 *
 * This script walks the repository, finds every quiz markdown file
 * (anything with "quiz" in the filename), and produces a JSON payload
 * that mirrors the Prisma schema shared in the request:
 *   - Quizz -> Question -> Attachment
 *   - Stores one .json file next to each .md file so existing image
 *     folders continue to work without moving assets around.
 *
 * Usage examples:
 *   node scripts/generate-quiz-json.js
 *   node scripts/generate-quiz-json.js --root . --output . --overwrite
 *   node scripts/generate-quiz-json.js --dry-run --match python
 *
 * Flags:
 *   --root <dir>       : folder to scan (default: cwd)
 *   --output <dir>     : base output folder (default: same as --root)
 *   --overwrite        : rewrite existing .json files instead of skipping
 *   --dry-run          : parse and report without writing files
 *   --match <keyword>  : only process files whose path includes the keyword
 *   --created-by <id>  : seed user id (default: SEED_USER_ID env or "seed-user")
 *
 * The JSON file shape is intentionally close to the Prisma models so it
 * can be consumed by a later Prisma seed script without further parsing:
 * {
 *   "quizz": {
 *     "id": "...",
 *     "title": "...",
 *     "description": "...",
 *     "createdById": "...",
 *     "questions": [
 *       {
 *         "id": "...",
 *         "question": "...",
 *         "options": ["..."],
 *         "correctAnswer": ["..."],
 *         "nature": "ChooseOne" | "ChooseMany",
 *         "answer": "...",
 *         "hint": "...",
 *         "explanation": "...",
 *         "attachments": [
 *           { "id": "...", "url": "...", "type": "question" }
 *         ]
 *       }
 *     ]
 *   },
 *   "meta": {
 *     "source": "python/python-quiz.md",
 *     "language": "fr",
 *     "generatedAt": "2024-01-01T00:00:00.000Z"
 *   }
 * }
 */

const crypto = require('crypto');
const fs = require('fs/promises');
const path = require('path');

async function main() {
  const config = parseArgs(process.argv.slice(2));
  const root = path.resolve(config.root ?? process.cwd());
  const outputRoot = path.resolve(config.output ?? root);

  const quizFiles = await findQuizFiles(root, config.match);
  if (!quizFiles.length) {
    console.warn('No quiz markdown files found.');
    return;
  }

  const summary = { written: 0, skipped: 0, warnings: 0, files: quizFiles.length };
  for (const filePath of quizFiles) {
    const relPath = path.relative(root, filePath);

    try {
      const content = await fs.readFile(filePath, 'utf8');
      const parsed = parseQuizMarkdown({
        content,
        sourcePath: relPath,
        createdById: config.createdById,
      });

      const outputFile = buildOutputPath({ root, outputRoot, sourcePath: relPath });
      if (!config.overwrite && !(config.dryRun)) {
        const exists = await fileExists(outputFile);
        if (exists) {
          summary.skipped += 1;
          console.info(`Skipping existing file: ${outputFile}`);
          continue;
        }
      }

      const serialized = JSON.stringify(parsed, null, 2);
      if (config.dryRun) {
        console.info(`Would write: ${outputFile}`);
      } else {
        await fs.mkdir(path.dirname(outputFile), { recursive: true });
        await fs.writeFile(outputFile, serialized, 'utf8');
        console.info(`Wrote ${outputFile} (${parsed.quizz.questions.length} questions)`);
      }

      summary.written += config.dryRun ? 0 : 1;
      summary.warnings += parsed.meta.warnings.length;
      for (const warn of parsed.meta.warnings) {
        console.warn(`[${relPath}] ${warn}`);
      }
    } catch (error) {
      summary.warnings += 1;
      console.error(`Failed to parse ${relPath}:`, error.message);
    }
  }

  console.info('-----');
  console.info(`Processed ${summary.files} files`);
  console.info(`Written: ${summary.written} | Skipped: ${summary.skipped} | Warnings: ${summary.warnings}`);
}

/**
 * Parse CLI arguments without external dependencies.
 */
function parseArgs(argv) {
  const config = {
    root: process.cwd(),
    output: null,
    overwrite: false,
    dryRun: false,
    match: null,
    createdById: process.env.SEED_USER_ID || 'seed-user',
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    switch (arg) {
      case '--root':
        config.root = argv[++i];
        break;
      case '--output':
        config.output = argv[++i];
        break;
      case '--overwrite':
        config.overwrite = true;
        break;
      case '--dry-run':
        config.dryRun = true;
        break;
      case '--match':
        config.match = argv[++i];
        break;
      case '--created-by':
        config.createdById = argv[++i];
        break;
      default:
        console.warn(`Unknown argument ignored: ${arg}`);
    }
  }

  return config;
}

/**
 * Recursively walk the repository to find quiz markdown files.
 * A quiz file is any .md with "quiz" in the filename.
 */
async function findQuizFiles(root, keyword) {
  const results = [];

  async function walk(dir) {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name.startsWith('.git')) continue;
      if (entry.name === 'node_modules') continue;

      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(fullPath);
      } else if (entry.isFile() && entry.name.endsWith('.md') && entry.name.toLowerCase().includes('quiz')) {
        if (keyword && !fullPath.includes(keyword)) continue;
        results.push(fullPath);
      }
    }
  }

  await walk(root);
  return results.sort();
}

/**
 * Parse a quiz markdown file into a JSON shape that mirrors the Prisma schema.
 */
function parseQuizMarkdown({ content, sourcePath, createdById }) {
  const warnings = [];
  const lines = content.split(/\r?\n/);

  let title = null;
  const introLines = [];
  const questionBlocks = [];
  let currentQuestion = null;
  let inQuestions = false;

  for (const line of lines) {
    if (!title && line.startsWith('## ')) {
      title = line.replace(/^##\s*/, '').trim();
      continue;
    }

    if (line.startsWith('#### ')) {
      inQuestions = true;
      if (currentQuestion) {
        questionBlocks.push(currentQuestion);
      }
      currentQuestion = {
        heading: line.replace(/^####\s*/, '').trim(),
        body: [],
      };
      continue;
    }

    if (!inQuestions) {
      introLines.push(line);
    } else if (currentQuestion) {
      currentQuestion.body.push(line);
    }
  }

  if (currentQuestion) {
    questionBlocks.push(currentQuestion);
  }

  const description = introLines.join(' ').replace(/\s+/g, ' ').trim() || `Seeded from ${sourcePath}`;
  const quizId = stableId('quiz', sourcePath);

  const questions = questionBlocks.map((block, index) => {
    const parsedQuestion = parseQuestionBlock({
      block,
      quizId,
      questionIndex: index,
    });

    if (!parsedQuestion.correctAnswer.length) {
      warnings.push(`Question ${index + 1} has no marked correct answers`);
    }
    if (!parsedQuestion.options.length) {
      warnings.push(`Question ${index + 1} has no options`);
    }
    return parsedQuestion;
  });

  return {
    quizz: {
      id: quizId,
      title: title || buildTitleFromPath(sourcePath),
      description,
      createdById,
      questions,
    },
    meta: {
      source: sourcePath,
      language: deriveLanguageFromFilename(sourcePath),
      generatedAt: new Date().toISOString(),
      warnings,
    },
  };
}

/**
 * Parse a single question block into a structured object.
 */
function parseQuestionBlock({ block, quizId, questionIndex }) {
  const optionPattern = /^\s*[-*+]\s*\[( |x|X)\]\s*(.+)$/;
  const questionLines = [];
  const trailingLines = [];
  const options = [];
  const correctAnswer = [];
  const attachments = [];
  let insideOptions = false;
  let insideCode = false;

  const baseQuestion = stripQuestionNumber(block.heading);

  for (const line of block.body) {
    if (line.trim().startsWith('```')) {
      insideCode = !insideCode;
    }

    // Collect attachments from inline markdown images.
    for (const match of line.matchAll(/!\[[^\]]*]\(([^)]+)\)/g)) {
      attachments.push({
        id: stableId('attachment', quizId, questionIndex, match[1]),
        url: match[1],
        type: 'question',
      });
    }

    const optionMatch = line.match(optionPattern);
    if (optionMatch && !insideCode) {
      insideOptions = true;
      const text = optionMatch[2].trim();
      options.push(text);
      if (optionMatch[1].toLowerCase() === 'x') {
        correctAnswer.push(text);
      }
      continue;
    }

    if (!insideOptions) {
      questionLines.push(line);
    } else {
      trailingLines.push(line);
    }
  }

  const hintLines = [];
  const explanationLines = [];
  for (const line of trailingLines) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    if (/^hint[:\-]?\s*/i.test(trimmed)) {
      hintLines.push(trimmed.replace(/^hint[:\-]?\s*/i, ''));
    } else {
      explanationLines.push(trimmed);
    }
  }

  const questionTextParts = [baseQuestion];
  const extraQuestionText = questionLines.join('\n').trim();
  if (extraQuestionText) {
    questionTextParts.push(extraQuestionText);
  }

  return {
    id: stableId('question', quizId, questionIndex, block.heading),
    question: questionTextParts.join('\n').trim(),
    answer: correctAnswer.join('; ') || null,
    explanation: explanationLines.join('\n').trim() || null,
    hint: hintLines.join('\n').trim() || null,
    correctAnswer,
    options,
    nature: correctAnswer.length > 1 ? 'ChooseMany' : 'ChooseOne',
    attachments,
  };
}

function stripQuestionNumber(text) {
  // Remove leading tokens like "Q1.", "P2.", "Pregunta 3.", etc.
  const match = text.match(/^(?:[A-Za-zÀ-ÿ?¿¡']*\s*)?\d+\.?\s*(.*)$/);
  if (match && match[1]) {
    return match[1].trim();
  }
  return text.trim();
}

function deriveLanguageFromFilename(filePath) {
  const base = path.basename(filePath);
  const match = base.match(/-quiz[-.]([a-z]{2}(?:-[A-Za-z]{2})?)/i);
  if (match) return match[1];
  return 'en';
}

function buildTitleFromPath(sourcePath) {
  const segments = sourcePath.split(path.sep);
  const folder = segments.length ? segments[0] : 'Quiz';
  return folder
    .split(/[-_]/)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ');
}

function buildOutputPath({ root, outputRoot, sourcePath }) {
  const baseDir = path.dirname(sourcePath);
  const baseName = path.basename(sourcePath, '.md');
  const targetDir = path.join(outputRoot, baseDir);
  return path.join(targetDir, `${baseName}.json`);
}

function stableId(...parts) {
  const hash = crypto.createHash('sha256');
  for (const part of parts) {
    hash.update(String(part || ''));
    hash.update('|');
  }
  return hash.digest('hex').slice(0, 24);
}

async function fileExists(filePath) {
  try {
    await fs.access(filePath);
    return true;
  } catch {
    return false;
  }
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: generate_quiz_json.py
#!/usr/bin/env python3
"""
Quiz JSON generator for Prisma seeding.

Walks the repository, finds markdown files with "quiz" in the name, and writes
JSON siblings that mirror the Prisma Quizz/Question/Attachment shape so they
can be consumed directly by a seed script.

Usage examples:
  python generate_quiz_json.py
  python generate_quiz_json.py --root . --output . --overwrite
  python generate_quiz_json.py --dry-run --match python
"""

import argparse
import hashlib
import json
import os
import re
from datetime import datetime, timezone
from pathlib import Path


def main() -> None:
    args = parse_args()
    root = Path(args.root).resolve()
    output_root = Path(args.output or root).resolve()
    created_by_id = args.created_by or os.getenv(
        "SEED_USER_ID", "cmiz68drf00004eqsc3izonqy")

    quiz_files = find_quiz_files(root, args.match)
    if not quiz_files:
        print("No quiz markdown files found.")
        return

    summary = {"files": len(quiz_files), "written": 0,
               "skipped": 0, "warnings": 0}
    for quiz_file in quiz_files:
        rel_path = quiz_file.relative_to(root)
        try:
            content = quiz_file.read_text(encoding="utf-8")
            parsed = parse_quiz_markdown(
                content=content, source_path=str(rel_path), created_by_id=created_by_id
            )
            output_file = build_output_path(
                root=root, output_root=output_root, source_path=str(rel_path)
            )

            if not args.overwrite and not args.dry_run and output_file.exists():
                summary["skipped"] += 1
                print(f"Skipping existing file: {output_file}")
                continue

            if args.dry_run:
                print(f"Would write: {output_file}")
            else:
                output_file.parent.mkdir(parents=True, exist_ok=True)
                output_file.write_text(json.dumps(
                    parsed, indent=2), encoding="utf-8")
                print(
                    f"Wrote {output_file} "
                    f"({len(parsed['quizz']['questions'])} questions)"
                )

            summary["written"] += 0 if args.dry_run else 1
            summary["warnings"] += len(parsed["meta"]["warnings"])
            for warn in parsed["meta"]["warnings"]:
                print(f"[{rel_path}] {warn}")
        except Exception as exc:  # pylint: disable=broad-except
            summary["warnings"] += 1
            print(f"Failed to parse {rel_path}: {exc}")

    print("-----")
    print(
        f"Processed {summary['files']} files | "
        f"Written: {summary['written']} | "
        f"Skipped: {summary['skipped']} | "
        f"Warnings: {summary['warnings']}"
    )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Generate JSON quiz payloads from markdown files."
    )
    parser.add_argument("--root", default=".", help="Root folder to scan.")
    parser.add_argument(
        "--output",
        default=None,
        help="Output folder base (defaults to the same as --root).",
    )
    parser.add_argument(
        "--overwrite", action="store_true", help="Rewrite existing JSON files."
    )
    parser.add_argument("--dry-run", action="store_true",
                        help="Do not write files.")
    parser.add_argument(
        "--match",
        default=None,
        help="Only process files whose path contains this substring.",
    )
    parser.add_argument(
        "--created-by",
        dest="created_by",
        default=None,
        help="createdById to set on quizzes (default: SEED_USER_ID or 'cmiz68drf00004eqsc3izonqy').",
    )
    return parser.parse_args()


def find_quiz_files(root: Path, keyword: str | None) -> list[Path]:
    results: list[Path] = []
    skip_dirs = {".git", "node_modules", ".next", ".turbo"}
    for path in root.rglob("*quiz*.md"):
        if any(part in skip_dirs for part in path.parts):
            continue
        if keyword and keyword not in str(path):
            continue
        results.append(path)
    return sorted(results)


def parse_quiz_markdown(content: str, source_path: str, created_by_id: str) -> dict:
    lines = content.splitlines()
    title = None
    intro_lines: list[str] = []
    question_blocks: list[dict] = []
    current_block: dict | None = None
    in_questions = False

    for line in lines:
        if title is None and line.startswith("## "):
            title = line.replace("##", "", 1).strip()
            continue

        if line.startswith("#### "):
            in_questions = True
            if current_block:
                question_blocks.append(current_block)
            current_block = {"heading": line.replace(
                "####", "", 1).strip(), "body": []}
            continue

        if not in_questions:
            intro_lines.append(line)
        elif current_block is not None:
            current_block["body"].append(line)

    if current_block:
        question_blocks.append(current_block)

    description = (
        " ".join(intro_lines).strip() or f"Seeded from {source_path}"
    )
    quizz_id = stable_id("quiz", source_path)

    questions = []
    warnings: list[str] = []

    for idx, block in enumerate(question_blocks):
        parsed = parse_question_block(block, quizz_id, idx)
        if not parsed["correctAnswer"]:
            warnings.append(
                f"Question {idx + 1} has no marked correct answers")
        if not parsed["options"]:
            warnings.append(f"Question {idx + 1} has no options")
        questions.append(parsed)

    return {
        "quizz": {
            "id": quizz_id,
            "title": title or build_title_from_path(source_path),
            "description": description,
            "createdById": created_by_id,
            "questions": questions,
        },
        "meta": {
            "source": source_path,
            "language": derive_language_from_filename(source_path),
            "generatedAt": datetime.now(timezone.utc).isoformat(),
            "warnings": warnings,
        },
    }


def parse_question_block(block: dict, quizz_id: str, question_index: int) -> dict:
    option_re = re.compile(r"^\s*[-*+]\s*\[( |x|X)\]\s*(.+)$")
    question_lines: list[str] = []
    trailing_lines: list[str] = []
    options: list[str] = []
    correct: list[str] = []
    attachments: list[dict] = []
    inside_options = False
    inside_code = False

    heading_text = strip_question_number(block["heading"])

    for line in block["body"]:
        if line.strip().startswith("```"):
            inside_code = not inside_code

        for match in re.finditer(r"!\[[^\]]*]\(([^)]+)\)", line):
            attachments.append(
                {
                    "id": stable_id(
                        "attachment", quizz_id, question_index, match.group(1)
                    ),
                    "url": match.group(1),
                    "type": "question",
                }
            )

        option_match = None if inside_code else option_re.match(line)
        if option_match:
            inside_options = True
            text = option_match.group(2).strip()
            options.append(text)
            if option_match.group(1).lower() == "x":
                correct.append(text)
            continue

        if not inside_options:
            question_lines.append(line)
        else:
            trailing_lines.append(line)

    hint_lines: list[str] = []
    explanation_lines: list[str] = []
    for line in trailing_lines:
        trimmed = line.strip()
        if not trimmed:
            continue
        if re.match(r"^hint[:\-]?\s*", trimmed, re.IGNORECASE):
            hint_lines.append(
                re.sub(r"^hint[:\-]?\s*", "", trimmed, flags=re.IGNORECASE))
        else:
            explanation_lines.append(trimmed)

    extra_question = "\n".join(q for q in question_lines if q.strip()).strip()
    question_text_parts = [heading_text]
    if extra_question:
        question_text_parts.append(extra_question)

    nature = "ChooseMany" if len(correct) > 1 else "ChooseOne"

    return {
        "id": stable_id("question", quizz_id, question_index, heading_text),
        "question": "\n".join(question_text_parts).strip(),
        "answer": "; ".join(correct) if correct else None,
        "explanation": "\n".join(explanation_lines).strip() or None,
        "hint": "\n".join(hint_lines).strip() or None,
        "correctAnswer": correct,
        "options": options,
        "nature": nature,
        "attachments": attachments,
    }


def strip_question_number(text: str) -> str:
    match = re.match(r"^(?:[A-Za-zÀ-ÿ?¿¡']*\s*)?\d+\.?\s*(.*)$", text.strip())
    return match.group(1).strip() if match and match.group(1) else text.strip()


def derive_language_from_filename(file_path: str) -> str:
    base = Path(file_path).name
    match = re.search(
        r"-quiz[-.]([a-z]{2}(?:-[A-Za-z]{2})?)", base, re.IGNORECASE)
    return match.group(1) if match else "en"


def build_title_from_path(source_path: str) -> str:
    folder = source_path.split(os.sep)[0]
    return " ".join(part.capitalize() for part in re.split(r"[-_]", folder))


def build_output_path(root: Path, output_root: Path, source_path: str) -> Path:
    rel_dir = Path(source_path).parent
    base_name = Path(source_path).stem
    target_dir = output_root / rel_dir
    return target_dir / f"{base_name}.json"


def stable_id(*parts: object) -> str:
    h = hashlib.sha256()
    for part in parts:
        h.update(str(part or "").encode("utf-8"))
        h.update(b"|")
    return h.hexdigest()[:24]


if __name__ == "__main__":
    main()
